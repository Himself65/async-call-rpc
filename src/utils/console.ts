//#region Console
/**
 * The minimal Console interface that AsyncCall needs.
 * @public
 */
export interface Console {
    debug?(...args: unknown[]): void
    log(...args: unknown[]): void
    groupCollapsed?(...args: unknown[]): void
    groupEnd?(...args: unknown[]): void
    error?(...args: unknown[]): void
}
