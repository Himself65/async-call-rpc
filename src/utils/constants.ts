export const isString = (x: unknown): x is string => typeof x === 'string'
export const isBoolean = (x: unknown): x is boolean => typeof x === 'boolean'
export const isFunction = (x: unknown): x is Function => typeof x === 'function'
export const isObject = (params: any): params is object => typeof params === 'object' && params !== null
export const ERROR = 'Error'
export const undefined = void 0
export const Object_setPrototypeOf = Object.setPrototypeOf
export const Promise_resolve = <T>(x: T) => Promise.resolve(x)
export const isArray = Array.isArray
export const replayFunction = () => '() => replay()'
