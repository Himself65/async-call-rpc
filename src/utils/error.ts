class CustomError extends Error {
    constructor(public name: string, message: string, public code: number, public stack: string) {
        super(message)
    }
}
/** These Error is defined in ECMAScript spec */
const errors: Record<string, typeof EvalError> = {
    Error,
    EvalError,
    RangeError,
    ReferenceError,
    SyntaxError,
    TypeError,
    URIError,
}
export const DOMExceptionHeader = 'DOMException:'
/**
 * AsyncCall support somehow transfer ECMAScript Error
 */
export function RecoverError(type: string, message: string, code: number, stack: string): Error {
    try {
        if (type.startsWith(DOMExceptionHeader) && DOMException) {
            const [, name] = type.split(DOMExceptionHeader)
            return new DOMException(message, name)
        } else if (type in errors) {
            const e = new errors[type](message)
            e.stack = stack
            Object.assign(e, { code })
            return e
        } else {
            return new CustomError(type, message, code, stack)
        }
    } catch {
        return new Error(`E${code} ${type}: ${message}\n${stack}`)
    }
}
export function removeStackHeader(stack = '') {
    return stack.replace(/^.+\n.+\n/, '')
}
function getDOMException(): { new (message: string, name: string): any } | undefined {
    const x = Reflect.get(globalThis, 'DOMException')
    if (typeof x === 'function') return x
    return undefined
}
export const DOMException = getDOMException()
