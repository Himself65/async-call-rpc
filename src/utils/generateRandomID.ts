export const generateRandomID = () => Math.random().toString(36).slice(2)
