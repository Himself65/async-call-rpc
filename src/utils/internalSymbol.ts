export const AsyncCallIgnoreResponse = Symbol.for('AsyncCall: This response should be ignored.')
