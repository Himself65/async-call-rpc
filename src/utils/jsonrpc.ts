import { DOMException, DOMExceptionHeader } from './error'

const jsonrpc = '2.0'
type ID = string | number | null | undefined

type JSONRPC = Readonly<{
    jsonrpc: typeof jsonrpc
    id: ID
}>
export type Request = Readonly<{
    method: string
    params: readonly unknown[] | object
}> & { remoteStack?: string } & JSONRPC
export function Request(id: ID, method: string, params: readonly unknown[] | object, remoteStack: string): Request {
    const x: Request = { jsonrpc, id, method, params, remoteStack }
    if (remoteStack.length === 0) delete x.remoteStack
    return x
}

export type SuccessResponse = Readonly<{ result: unknown }> & JSONRPC
export function SuccessResponse(id: ID, result: any): SuccessResponse {
    const x: SuccessResponse = { jsonrpc, id, result }
    return x
}

export type ErrorResponse = JSONRPC &
    Readonly<{
        error: { code: number; message: string; data?: { stack?: string; type?: string } }
    }>

export function ErrorResponse(id: ID, code: number, message: string, stack: string, e?: unknown): ErrorResponse {
    let type = toString('Error', () => (e as any)?.constructor?.name)
    if (DOMException && e instanceof DOMException) type = DOMExceptionHeader + e.name
    if (typeof e === 'string' || typeof e === 'number' || typeof e === 'boolean' || typeof e === 'bigint') {
        type = 'Error'
        message = String(e)
    }
    if (id === undefined) id = null
    code = Math.floor(code)
    if (Number.isNaN(code)) code = -1
    const error: ErrorResponse['error'] = { code, message, data: { stack, type } }
    return { error, id, jsonrpc }
}

// Pre defined error in section 5.1
ErrorResponse.ParseError = (stack = '') => ErrorResponse(null, -32700, 'Parse error', stack)
ErrorResponse.InvalidRequest = (id: ID) => ErrorResponse(id, -32600, 'Invalid Request', '')
ErrorResponse.MethodNotFound = (id: ID) => ErrorResponse(id, -32601, 'Method not found', '')
ErrorResponse.InvalidParams = (id: ID) => ErrorResponse(id, -32602, 'Invalid params', '')
ErrorResponse.InternalError = (id: ID, message: string = '') =>
    ErrorResponse(id, -32603, 'Internal error' + message, '')

export type Response = SuccessResponse | ErrorResponse

export function isJSONRPCObject(data: any): data is Response | Request {
    if (!isObject(data)) return false
    if (!hasKey(data, 'jsonrpc')) return false
    if (data.jsonrpc !== jsonrpc) return false
    if (hasKey(data, 'params')) {
        const params = (data as Request).params
        if (!Array.isArray(params) && !isObject(params)) return false
    }
    return true
}

export function isObject(params: any): params is object {
    return typeof params === 'object' && params !== null
}

export function hasKey<T, Q extends string>(
    obj: T,
    key: Q,
): obj is T &
    {
        [key in Q]: unknown
    } {
    return key in obj
}

function toString(def: string, val: () => any) {
    let str = def
    try {
        str = val()
    } catch {}
    if (typeof str !== 'string') return def
    return str
}
