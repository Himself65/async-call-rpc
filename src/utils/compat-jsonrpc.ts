import { ErrorResponse, Request, SuccessResponse } from './jsonrpc'

export type ID = string | number | null | undefined
type CompatRequest = [type: 0, method: string, params: readonly unknown[] | object, id?: ID, stack?: string]
// [normal: 0, id, method, ...params]
// [+stack: 1, id, method, stack, ...params]
// [notify: 2, method, ...params]
// [+stack: 3, stack, ...params]
type CompatSuccessResponse = [type: 1, id: ID, result?: unknown]
type CompatErrorResponse = [type: 2, id: ID, code: number, message: string, data?: unknown]
export const CompressRequest = (req: Request): CompatRequest => {
    const x: CompatRequest = [0, req.method, req.params]
    if (req.id !== undefined) x[3] = req.id
    if (req.remoteStack) x[4] = req.remoteStack
    return x
}
export const DecompressRequest = ([, methods, param, id, stack]: CompatRequest) => Request(id, methods, param, stack)
export const CompressSuccessResponse = ({ result, id }: SuccessResponse): CompatSuccessResponse => {
    const x: CompatSuccessResponse = [1, id]
    if (result !== undefined) x[2] = result
    return x
}
export const DecompressSuccessResponse = ([, ...r]: CompatSuccessResponse) => SuccessResponse(...r)
export const CompressErrorResponse = ({ error: { code, message, data }, id }: ErrorResponse): CompatErrorResponse => {
    const x: CompatErrorResponse = [2, id, code, message]
    if (data !== undefined) x[4] = data
    return x
}
export const DecompressErrorResponse = ([, ...y]: CompatErrorResponse): ErrorResponse => ErrorResponse(...y)
export const Decompress = (x: CompatRequest | CompatErrorResponse | CompatSuccessResponse) => {
    if (x[0] == 0) return DecompressRequest(x)
    if (x[0] == 1) return DecompressSuccessResponse(x)
    return DecompressErrorResponse(x)
}
