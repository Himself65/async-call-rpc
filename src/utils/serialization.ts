//#region Serialization

import { isObject, hasKey } from './jsonrpc'

/**
 * Serialization and deserialization of the JSON RPC payload
 * @public
 */
export interface Serialization {
    /**
     * Do serialization
     * @param from - original data
     */
    serialization(from: any): unknown | PromiseLike<unknown>
    /**
     * Do deserialization
     * @param serialized - Serialized data
     */
    deserialization(serialized: unknown): unknown | PromiseLike<unknown>
}

/**
 * Serialization implementation that do nothing
 * @remarks {@link Serialization}
 * @public
 */
export const NoSerialization: Serialization = {
    serialization(from) {
        return from
    },
    deserialization(serialized) {
        return serialized
    },
}

/**
 * Create a serialization by JSON.parse/stringify
 *
 * @param replacerAndReceiver - Replacer and receiver of JSON.parse/stringify
 * @param space - Adds indentation, white space, and line break characters to the return-value JSON text to make it easier to read.
 * @param undefinedKeepingBehavior - How to keep "undefined" in result of SuccessResponse?
 *
 * If it is not handled properly, JSON.stringify will emit an invalid JSON RPC object.
 *
 * Options:
 * - `"null"`(**default**): convert it to null.
 * - `"keep"`: try to keep it by additional property "undef".
 * - `false`: Don't keep it, let it break.
 * @remarks {@link Serialization}
 * @public
 */
export const JSONSerialization = (
    replacerAndReceiver: [((key: string, value: any) => any)?, ((key: string, value: any) => any)?] = [
        undefined,
        undefined,
    ],
    space?: string | number | undefined,
    undefinedKeepingBehavior: 'keep' | 'null' | false = 'null',
): Serialization => ({
    serialization(from) {
        if (undefinedKeepingBehavior && isObject(from) && hasKey(from, 'result') && from.result === undefined) {
            const alt = Object.assign({}, from)
            alt.result = null
            if (undefinedKeepingBehavior === 'keep') (alt as any).undef = true
            from = alt
        }
        return JSON.stringify(from, replacerAndReceiver[0], space)
    },
    deserialization(serialized) {
        const result = JSON.parse(serialized as string, replacerAndReceiver[1])
        if (
            isObject(result) &&
            hasKey(result, 'result') &&
            result.result === null &&
            hasKey(result, 'undef') &&
            result.undef === true
        ) {
            result.result = undefined
            delete result.undef
        }
        return result
    },
})
//#endregion
