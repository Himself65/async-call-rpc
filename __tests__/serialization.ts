import { NoSerialization, JSONSerialization } from '../src/utils/serialization'
import { SuccessResponse } from '../src/utils/jsonrpc'

test('NoSerialization', () => {
    const x = Symbol()
    expect(NoSerialization.serialization(x)).toBe(x)
    expect(NoSerialization.deserialization(x)).toBe(x)
})

test('JSONSerialization', () => {
    const y = { rand: 123, y: true, z: undefined }
    const json = JSONSerialization()
    const result = json.serialization(y)
    expect(result).toMatchSnapshot()
    expect(json.deserialization(result)).not.toStrictEqual(y)
    delete y.z
    expect(json.deserialization(result)).toStrictEqual(y)
})

test('JSONSerialization undefined special handling', () => {
    const y = SuccessResponse(1, undefined)
    const test = (behavior: Parameters<typeof JSONSerialization>[2]) => {
        const json = JSONSerialization(undefined, undefined, behavior)
        const s = json.serialization(y)
        expect(s).toMatchSnapshot(String(behavior))
        expect(json.deserialization(s)).toMatchSnapshot(String(behavior))
    }
    test(undefined)
    test(false)
    test('keep')
    test('null')
})

test('JSONSerialization with replacer', () => {
    const y = { foundation: 'Mozilla', model: 'box', week: 45, transport: 'car', month: 7 }
    const json = JSONSerialization([replacer, (key, value) => (typeof value === 'number' ? value * 2 : value)], 4)
    const result = json.serialization(y)
    expect(result).toMatchSnapshot()
    expect(json.deserialization(result)).not.toStrictEqual(y)
    expect(json.deserialization(result)).toMatchSnapshot()
})

function replacer(key: any, value: any) {
    if (typeof value === 'string') {
        return undefined
    }
    return value
}
