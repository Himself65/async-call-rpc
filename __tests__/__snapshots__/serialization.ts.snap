// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSerialization 1`] = `"{\\"rand\\":123,\\"y\\":true}"`;

exports[`JSONSerialization undefined special handling: false 1`] = `"{\\"jsonrpc\\":\\"2.0\\",\\"id\\":1}"`;

exports[`JSONSerialization undefined special handling: false 2`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
}
`;

exports[`JSONSerialization undefined special handling: keep 1`] = `"{\\"jsonrpc\\":\\"2.0\\",\\"id\\":1,\\"result\\":null,\\"undef\\":true}"`;

exports[`JSONSerialization undefined special handling: keep 2`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": undefined,
}
`;

exports[`JSONSerialization undefined special handling: null 1`] = `"{\\"jsonrpc\\":\\"2.0\\",\\"id\\":1,\\"result\\":null}"`;

exports[`JSONSerialization undefined special handling: null 2`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": null,
}
`;

exports[`JSONSerialization undefined special handling: undefined 1`] = `"{\\"jsonrpc\\":\\"2.0\\",\\"id\\":1,\\"result\\":null}"`;

exports[`JSONSerialization undefined special handling: undefined 2`] = `
Object {
  "id": 1,
  "jsonrpc": "2.0",
  "result": null,
}
`;

exports[`JSONSerialization with replacer 1`] = `
"{
    \\"week\\": 45,
    \\"month\\": 7
}"
`;

exports[`JSONSerialization with replacer 2`] = `
Object {
  "month": 14,
  "week": 90,
}
`;
